Approx repr ugly with complex numbers

MWE:

```python
def test_complex():
    assert pytest.approx(1j) == 2j
```

Result:

```

=================================== FAILURES ===================================
_________________________________ test_complex _________________________________

>   ???
E   assert <[TypeError("can't convert complex to float") raised in repr()] SafeRepr object at 0x7f40613577a0> == 2j
E    +  where <[TypeError("can't convert complex to float") raised in repr()] SafeRepr object at 0x7f40613577a0> = <class '_pytest.python.approx'>(1j)
E    +    where <class '_pytest.python.approx'> = pytest.approx
```

This should clearly be fixed.

Version:
```
>>> pytest.__version__
'3.0.3'
```

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -7,6 +7,10 @@
   subtle bugs (`#2078`_).
   Thanks `@nicoddemus`_ for the PR.
 
+* Fix error message using ``approx`` with complex numbers (`#2082`_).
+  Thanks `@adler-j`_ for the report and `@nicoddemus`_ for the PR.
+
+*
 
 * Cope gracefully with a .pyc file with no matching .py file (`#2038`_). Thanks
   `@nedbat`_.
@@ -15,10 +19,12 @@
 
 *
 
+.. _@adler-j: https://github.com/adler-j
 .. _@nedbat: https://github.com/nedbat
 
 .. _#2038: https://github.com/pytest-dev/pytest/issues/2038
 .. _#2078: https://github.com/pytest-dev/pytest/issues/2078
+.. _#2082: https://github.com/pytest-dev/pytest/issues/2082
 
 
 3.0.4

--- a/_pytest/python.py
+++ b/_pytest/python.py
@@ -1419,6 +1419,9 @@ class ApproxNonIterable(object):
         self.rel = rel
 
     def __repr__(self):
+        if isinstance(self.expected, complex):
+            return str(self.expected)
+
         # Infinities aren't compared using tolerances, so don't show a
         # tolerance.
         if math.isinf(self.expected):

--- a/testing/python/approx.py
+++ b/testing/python/approx.py
@@ -28,6 +28,7 @@ class TestApprox:
         print(approx(inf))
         print(approx(1.0, rel=nan))
         print(approx(1.0, rel=inf))
+        print(approx(1.0j, rel=inf))
 
     def test_operator_overloading(self):
         assert 1 == approx(1, rel=1e-6, abs=1e-12)

